File: TPII15_09/GoF/Criacionais/Singleton/Singleton02.java
import java.util.ArrayList;
import java.util.List;

public class Carrinho {
    private static Carrinho instance;
    private List<Produto> produtos;

    private Carrinho() {
        produtos = new ArrayList<>();
    }

    public static Carrinho getInstance() {
        if (instance == null) {
            instance = new Carrinho();
        }
        return instance;
    }

    public void addProduto(Produto produto) {
        produtos.add(produto);
    }

    public List<Produto> getProdutos() {
        return produtos;
    }

    public void clearCarrinho() {
        produtos.clear();
    }

    public static void main(String[] args) {
        // EXEMPLO DE USO ##############################################################
        Carrinho carrinhoInstance1 = Carrinho.getInstance();
        Carrinho carrinhoInstance2 = Carrinho.getInstance();

        System.out.println(carrinhoInstance1 == carrinhoInstance2);

        carrinhoInstance1.addProduto(new Produto(1, "Produto 1", 10.00));
        carrinhoInstance2.addProduto(new Produto(2, "Produto 2", 20.00));

        System.out.println(carrinhoInstance1.getProdutos());
        System.out.println(carrinhoInstance2.getProdutos());

        carrinhoInstance1.clearCarrinho();

        System.out.println(carrinhoInstance1.getProdutos());
        System.out.println(carrinhoInstance2.getProdutos());
    }
}

class Produto {
    private int id;
    private String nome;
    private double preco;

    public Produto(int id, String nome, double preco) {
        this.id = id;
        this.nome = nome;
        this.preco = preco;
    }

    @Override
    public String toString() {
        return "ID: " + id + ", Nome: " + nome + ", Preço: " + preco;
    }
}

File: TPII15_09/GoF/Criacionais/Singleton/Singleton.02.py
class Carrinho:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Carrinho, cls).__new__(cls)
            cls._instance.produtos = []
        return cls._instance

    def add_produto(self, produto):
        self.produtos.append(produto)

    def get_produtos(self):
        return self.produtos

    def clear_carrinho(self):
        self.produtos = []

# Classe Produto
class Produto:
    def __init__(self, id, nome, preco):
        self.id = id
        self.nome = nome
        self.preco = preco

    def __str__(self):
        return f"ID: {self.id}, Nome: {self.nome}, Preço: {self.preco}"

# EXEMPLO DE USO ##############################################################
carrinho_instance1 = Carrinho()
carrinho_instance2 = Carrinho()

print(carrinho_instance1 is carrinho_instance2)

produto1 = Produto(1, "Produto 1", 10.00)
produto2 = Produto(2, "Produto 2", 20.00)

carrinho_instance1.add_produto(produto1)
carrinho_instance2.add_produto(produto2)

print(carrinho_instance1.get_produtos())
print(carrinho_instance2.get_produtos())

carrinho_instance1.clear_carrinho()

print(carrinho_instance1.get_produtos())
print(carrinho_instance2.get_produtos())
